elixir-fundamentals-review.livemd

```diff
@@ -0,0 +1,339 @@
+
+  iex> Enumeration.evens([2, 2, 5, 4])
+  [2, 2, 4]
+  """
+  def evens(list) do
+    Enum.filter(list, fn each -> rem(each, 2) == 0 end)
+  end
+
+  @doc """
+  Sum only the even numbers in a list or range.
+
+  iex> Enumeration.sum_evens([2, 2, 5, 4])
+  8
+  """
+  def sum_evens(list) do
+    evens(list)
+    |> Enum.sum()
+  end
+end
+
+Enumeration.double(1..5)
+Enumeration.evens([2, 2, 5, 4])
+Enumeration.sum_evens([2, 2, 5, 4])
+```
+
+## Built In Modules
+
+You're going to use the built-in modules to perform the common CRUD (create, read, update, destroy/delete) operations on each data structure.
+
+<!-- livebook:{"break_markdown":true} -->
+
+### Maps
+
+<!-- livebook:{"break_markdown":true} -->
+
+Update (increment) the `:count` key in this map to return `%{count: 2}`.
+
+```elixir
+old_map = %{count: 1}
+Map.update(old_map, :count, 0, fn each -> each + 1 end)
+```
+
+Set the `:greeting` key in this map to be `"hi"` instead of `"hello"`.
+
+```elixir
+old_map = %{greeting: "hello"}
+Map.put(old_map, :greeting, "hi")
+```
+
+Delete the `:complete` key from this map to return an empty map.
+
+```elixir
+old_map = %{complete: "task"}
+Map.delete(old_map, :complete)
+```
+
+### List
+
+<!-- livebook:{"break_markdown":true} -->
+
+Delete the `"a"` element in this list.
+
+```elixir
+List.delete(["a", "b", "c"], "a")
+```
+
+Insert a `"b"` element at index `1` in this list.
+
+```elixir
+List.insert_at(["a", "c"], 1, "b")
+```
+
+Update the element at index `1` in this list to be lowercase.
+
+```elixir
+List.update_at(["a", "B", "c"], 1, fn each -> String.downcase(each) end)
+```
+
+### Keyword List
+
+<!-- livebook:{"break_markdown":true} -->
+
+Delete the `:a` key/value pair in this keyword list.
+
+```elixir
+Keyword.delete([a: 1], :a)
+```
+
+Insert a `:b` key with the value `2` in this keyword list at index `1` to create `[a: 1, b: 2, c: 3]`. Order does matter.
+
+<details style="background-color: burlywood; padding: 1rem; margin: 1rem 0;">
+<summary>Hint</summary>
+
+Remember that you can use `List` functions such as `List.insert_at/3` with keyword lists.
+
+</details>
+
+```elixir
+List.insert_at([a: 1, c: 3], 1, b: 2)
+```
+
+Retrieve the value for the `:label` key in this keyword list of options.
+
+```elixir
+Keyword.get([label: "some label"], :label)
+```
+
+## Strings and Regex
+
+Split all of the letters in this string to make `["a", "b", "c"]`.
+
+```elixir
+String.split("abc", "", trim: true)
+```
+
+Find all of the single digit characters in this string to make `["1", "2", "2", "3"]`
+
+```elixir
+"a1b22c3"
+```
+
+Determine if the `<<2::2>>` bitstring is a valid binary. This should be `false`.
+
+```elixir
+<<2::2>>
+```
+
+Determine if `<<2::2>>` is a bitstring. This should be `true`.
+
+```elixir
+<<2::2>>
+```
+
+Determine if `"hello"` is a valid binary. This should be `true`.
+
+```elixir
+"hello"
+```
+
+## Recursion
+
+Use recursion to sum every value in a list.
+
+```elixir
+defmodule Recursion do
```

Comment:
There are a few solutions.

We can use body recursion.

```elixir
def sum([]), do: 0
def sum([head | tail]), do: head + sum(tail)
```

Or tail recursion.

```elixir
def sum(_list, acc \\ 0)
def sum([], acc), do: acc
def sum([head | tail], acc) do
   sum(tail, head + acc)
end
```